### PR TITLE
refactor: introduce typed message & hint

### PR DESCRIPTION
**Pull request by magurotuna**
As discussed in #457, I gave it a shot to introduce typed message & hint with `derive_more` crate for the following three rules:

- eqeqeq
- no-delete-var
- getter-return

This looks semantically better to me. What do you think? @bartlomieju

### CODE REVIEW

File: src/linter.rs [hunk 1]
```diff
@@ -43,30 +43,35 @@ impl Context {
...
     self.diagnostics.push(diagnostic);
   }
 
   fn create_diagnostic(
     &self,
     span: Span,
-    code: impl Into<String>,
-    message: impl Into<String>,
+    code: String,
+    message: String,
```
Comment (magurotuna) [inline src/linter.rs:74]: This refactor is inspired by https://www.possiblerust.com/pattern/non-generic-inner-functions, in which the author says it is better to decrease generic functons as few as possible for faster compile time and less binary size.

File: src/rules/getter_return.rs [hunk 2]
```diff
@@ -386,134 +401,129 @@ const obj = {
...
       GetterReturn,
 
       // object getter
       "const foo = { get getter() {} };": [
         {
           col: 14,
-          message: msg1("getter"),
-          hint: "Return a value from the getter function",
+          message: GetterReturnMessage::Expected("getter".to_string()),
+          hint: GetterReturnHint::Return,
```
Comment (magurotuna) [inline src/rules/getter_return.rs:412]: `.to_string()` may look noisy here. I tried to remove it by some macro, but I couldn't. Probably it requres proc macro.

File: src/linter.rs (hunk 1 above)
Comment (bartlomieju) [reply src/linter.rs:74]: I'm pretty sure this won't matter much as most of the time is spent compiling swc and its macros anyway :)

Comment (magurotuna) [reply src/linter.rs:74]: Sounds right, to be honest I just wanted to practice it, so I don't have a strong will to do it. I will revert :)

File: src/rules/getter_return.rs (hunk 2 above)
Comment (bartlomieju) [reply src/rules/getter_return.rs:412]: The solution looks great, but those `to_string()` calls are noisy. @magurotuna maybe we could use `&str` instead with appropriate lifetime?

Comment (magurotuna) [reply src/rules/getter_return.rs:412]: sure, I'll give it a shot.

Comment (magurotuna) [reply src/rules/getter_return.rs:412]: hmmm... I'm doing my best but using `&str` in GetterRetrunMessage seems extremely difficult for me :/ (How I tried is here: 7759d4b810737ef3086f036e0f154fc0e307eb73)
I'll try another solution using a declarative macro.

File: src/rules/no_delete_var.rs [hunk 3]
```diff
@@ -101,7 +112,13 @@ mod tests {
   fn no_delete_var_invalid() {
     assert_lint_err! {
       NoDeleteVar,
-      r#"var someVar = "someVar"; delete someVar;"#: [{ col: 25, message: MESSAGE, hint: HINT }],
+      r#"var someVar = "someVar"; delete someVar;"#: [
+        {
+          col: 25,
+          message: variant!(NoDeleteVarMessage, Unexpected),
+          hint: variant!(NoDeleteVarHint, Remove),
```
Comment (bartlomieju) [inline src/rules/no_delete_var.rs:119]: This seems superfluous - `NoDeleteVarHint::Remove` is shorter 😃

File: src/rules/getter_return.rs [hunk 4]
```diff
@@ -97,35 +114,33 @@ impl<'c> GetterReturnVisitor<'c> {
...
     self.errors.insert(
       span,
-      format!(
-        "Expected to return a value in '{}'.",
+      GetterReturnMessage::Expected(
         self
           .getter_name
           .clone()
-          .expect("the name of getter is not set")
+          .expect("the name of getter is not set"),
```
Comment (bartlomieju) [inline src/rules/getter_return.rs:131]: We could potentially write this one as:
```rust
variant!(GetterReturnMessage, Expected, self.getter_name.unwrap())
```

File: src/rules/no_delete_var.rs (hunk 3 above)
Comment (magurotuna) [reply src/rules/no_delete_var.rs:119]: I was wondering which to choose, `variant!(FooEnum, Bar)` or `FooEnum::Bar` if the variant has no value. I don't have strong opinion about this, so let's go with your idea :) I'll rewrite soon.

File: src/rules/getter_return.rs (hunk 4 above)
Comment (magurotuna) [reply src/rules/getter_return.rs:131]: Currently `variant!` macro is in `test_util.rs`, which is compiled only when in tests due to `#[cfg(test)]`. 
Do you think `variant!` should be able to be used outside tests?

Comment (bartlomieju) [reply src/rules/getter_return.rs:131]: Let's leave it in `test_util.rs` for a bit and see if it will be useful in other rules 👍

Comment (magurotuna) [reply src/rules/getter_return.rs:131]: OK I see.

File: src/rules/no_delete_var.rs (hunk 3 above)
Comment (magurotuna) [reply src/rules/no_delete_var.rs]: Done

**Review by bartlomieju — APPROVED** [review]
LGTM, I agree this is cleaner. Let's land it and see impact on `deno` binary size in a few days.

**Pull request merged**